clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -314,69 +370,118 @@ public void poll(long timeout) throws IOException {
+                if (idleExpiryManager != null)
+                    idleExpiryManager.update(channel.id(), currentTimeNanos);
+
+                try {
+
+                    /* complete any connections that have finished their handshake (either normally or immediately) */
+                    if (isImmediatelyConnected || key.isConnectable()) {
+                        madeProgress = true;
+                        if (channel.finishConnect()) {
+                            this.connected.add(channel.id());
+                            this.sensors.connectionCreated.record();
+                            SocketChannel socketChannel = (SocketChannel) key.channel();
+                            log.debug("Created socket with SO_RCVBUF = {}, SO_SNDBUF = {}, SO_TIMEOUT = {} to node {}",
+                                    socketChannel.socket().getReceiveBufferSize(),
+                                    socketChannel.socket().getSendBufferSize(),
+                                    socketChannel.socket().getSoTimeout(),
+                                    channel.id());
+                        } else
+                            continue;
+                    }
 
-                /* complete any connections that have finished their handshake (either normally or immediately) */
-                if (isImmediatelyConnected || key.isConnectable()) {
-                    if (channel.finishConnect()) {
-                        this.connected.add(channel.id());
-                        this.sensors.connectionCreated.record();
-                        SocketChannel socketChannel = (SocketChannel) key.channel();
-                        log.debug("Created socket with SO_RCVBUF = {}, SO_SNDBUF = {}, SO_TIMEOUT = {} to node {}",
-                                socketChannel.socket().getReceiveBufferSize(),
-                                socketChannel.socket().getSendBufferSize(),
-                                socketChannel.socket().getSoTimeout(),
-                                channel.id());
-                    } else
-                        continue;
-                }
+                    /* if channel is not ready finish prepare */
+                    if (channel.isConnected() && !channel.ready()) {
+                        madeProgress = true;
+                        channel.prepare();
+                    }
 
-                /* if channel is not ready finish prepare */
-                if (channel.isConnected() && !channel.ready())
-                    channel.prepare();
+                    /* if channel is ready and has bytes to read from socket or buffer, and has no previous
+                       message(s) already staged or otherwise in progress then read from it */
+                    if (channel.ready() && (key.isReadable() || channel.hasBytesBuffered()) && !hasStagedReceive(channel)
+                            && !explicitlyMutedChannels.contains(channel)) {
```

Comment:
Is this test needed? If a channel is explicitly muted, it won't be selected by the selector, right?